src/lambdas/deleteUser/deleteUser.js

```diff
@@ -0,0 +1,50 @@
+const DynamoDB = require('aws-sdk/clients/dynamodb');
+const db = new DynamoDB.DocumentClient();
+const clone = require('clone');
+const Log = require('@dazn/lambda-powertools-logger');
+const TABLE_NAME = process.env.TABLE_NAME || '';
+
+module.exports.handler = async event => {
+    try {
+        const { id: userId } = event.pathParameters || {};
+        if (!userId) throw 'no user id';
+
+        // get item to delete
+        const originalParams = { TableName: TABLE_NAME, Key: { pk: userId, sk: `USER` } };
+        const doc = await db.get(originalParams).promise();
+        const { Item } = doc;
+        if (!Item) throw 'user does not exist';
+
+        // clone a backup item
+        const backupItem = clone(Item);
+        backupItem.sk = 'DELETED_USER';
+        backupItem.updatedAt = new Date().toISOString();
+
+        // store the backup
+        const backupParams = { TableName: TABLE_NAME, Item: backupItem };
+        await db.put(backupParams).promise();
+
+        //  delete the original
+        await db.delete(originalParams).promise();
+
+        // cloudwatch log
+        Log.info('User Deleted', { userId });
+
+        // respond success
+        return httpResponse(204);
+    } catch (err) {
+        Log.error('User Delete Error', new Error(err));
+        return httpResponse(400, err);
+    }
+};
+
+const httpResponse = (statusCode, body) => {
```

Comment:
Since this is common code repeated within lambdas, how would you recommend abstracting it to shared tooling?

Reply:
Normally I would put code like this in a shared module and import it into each lambda.  However, the CDK only supports deploying the lambda by specifying a folder or a ZIP file.  So putting it in a common module would have required duplicating the module in each lambda's directory, or putting all lambda code in the same directory, which creates lambdas with unused code from all the other lambdas.  Neither of which was ideal.  Because of this, and since the HTTP response function was small, I simply chose to duplicate it in each lambda.   

Eventually the AWS CDK is going to support efficiently bundling lambda code during synthesis of the CloudFormation ([see here](https://github.com/aws/aws-cdk/pull/5532)).  Until then, I would use webpack or parcel to to solve this, or write a build script that creates a ZIP package for each Lambda as a pre-deploy step.